File: typescript/src/lab1.ts
/**
 * This file compiles the code in Web Browser Engineering,
 * up to and including Chapter 1 (Downloading Web Pages)
 */

import assert from 'assert';
import net from 'net';
import { platform } from 'os';
import tls from 'tls';

let redirectsTimes = 0;

async function request(url: string) {
  const [scheme, ...other] = url.split('://');
  url = other.join('://');

  assert.ok(['http', 'https'].includes(scheme), `Unknown scheme ${scheme}`);

  let host: string;
  let path: string;
  if (url.includes('/')) {
    const res = url.split('/');
    host = res[0];
    path = `/${res.slice(1).join('/')}`;
  } else {
    host = url;
    path = '/';
  }

  let port = scheme === 'http' ? 80 : 443;

  if (host.includes(':')) {
    const res = host.split(':');
    host = res[0];
    port = Number(res.slice(1).join(':'));
  }

  let s = net.connect({
    host,
    port,
  });

  if (scheme === 'https') {
    s = tls.connect({
      socket: s,
      servername: host,
    });
  }

  s.once('connect', (stream) => {
    s.write(`GET ${path} HTTP/1.0\r\n`);
    s.write(`Host: ${host}\r\n`);
    s.write('Connection: close\r\n');
    s.write(`User-Agent: Mozilla/5.0 (${platform()})\r\n`);
    s.write('\r\n');
    // s.end('\r\n');
  });
  s.on('close', () => {
    console.log('socket close')
  })
  s.on('error', err => {
    console.log('socket error:', err)
  });

  const response = await new Promise<string>((resolve) => {
    let resp = '';
    s.on('data', (data) => {
      resp += data.toString('utf8')
    });
    s.on('end', () => {
      resolve(resp)
    })
  });

  const [statusLine, ...otherLines] = response.split('\r\n');
  const statusLineSplit = statusLine.split(' ');
  const status = statusLineSplit[1];
  const explanation = statusLineSplit.slice(2).join(' ');
  assert.ok(['200', '301', '302'].includes(status), `${status}: ${explanation}`);

  const headers: any = {};
  const firstEmptyLineIndex = otherLines.findIndex((e) => !e);
  const headersLines = otherLines.slice(0, firstEmptyLineIndex);
  const bodyLines = otherLines.slice(firstEmptyLineIndex + 1);

  headersLines.forEach((line) => {
    const res = line.split(':');
    const header = res[0];
    const value = res.slice(1).join(':');
    headers[header.toLowerCase()] = value.trim();
  });

  if (['301', '302'].includes(status)) {
    if (redirectsTimes > 10) {
      throw new Error('redirects times > 10 !')
    }
    if (!headers.location) {
      throw new Error('no headers.location')
    }
    redirectsTimes += 1;
    return request(headers.location)
  } else {
    redirectsTimes = 0;
  }

  const body = bodyLines.join('\r\n');
  // s.destroy();
  return {
    headers,
    body,
  };
}

class Text {
  text: string;

  constructor(text: string) {
    this.text = text;
  }
}

class Tag {
  tag: string;

  constructor(tag: string) {
    this.tag = tag;
  }
}

function show(body: string) {
  let inTag = false;
  let out = [];
  let text = '';
  for (const char of body) {
    if (char === '<') {
      inTag = true;
      if (text) {
        out.push(new Text(text));
      }
      text = '';
    } else if (char === '>') {
      inTag = false;
      out.push(new Tag(text));
      text = '';
    } else {
      text += char;
    }
  }

  if (!inTag && text) {
    out.push(new Text(text));
  }

  let inBody = false;
  let bodyText = ''

  for (const token of out) {
    if (token instanceof Text) {
      if (inBody) {
        bodyText += token.text;
      }
    } else if (token.tag === 'body') {
      inBody = true;
    } else if (token.tag === '/body') {
      inBody = false;
    }
  }

  console.log(bodyText);
}

async function load(url: string) {
  const { headers, body } = await request(url);
  show(body);
}

function main() {
  const url = process.argv[2];
  load(url);
}

main();

File: python/lab1.py
"""
This file compiles the code in Web Browser Engineering,
up to and including Chapter 1 (Downloading Web Pages)
"""

import socket
import ssl
import sys

redirects_times = 0

def request(url):
    scheme, url = url.split("://", 1)
    assert scheme in ["http", "https"], \
        "Unknown scheme {}".format(scheme)

    if ("/" in url):
        host, path = url.split("/", 1)
        path = "/" + path
    else:
        host = url
        path = '/'
    port = 80 if scheme == "http" else 443

    if ":" in host:
        host, port = host.split(":", 1)
        port = int(port)

    s = socket.socket(
        family=socket.AF_INET,
        type=socket.SOCK_STREAM,
        proto=socket.IPPROTO_TCP,
    )
    s.connect((host, port))

    if scheme == "https":
        ctx = ssl.create_default_context()
        s = ctx.wrap_socket(s, server_hostname=host)

    s.send("GET {} HTTP/1.0\r\n".format(path).encode("utf8"))
    s.send(("Host: {}\r\n".format(host)).encode("utf8"))
    s.send(("Connection: close\r\n").encode("utf8"))
    s.send(("User-Agent: Mozilla/5.0 ({sys.platform})\r\n").encode("utf8"))
    s.send(("\r\n").encode("utf8"))
    response = s.makefile("r", encoding="utf8", newline="\r\n")

    statusline = response.readline()
    version, status, explanation = statusline.split(" ", 2)
    assert status in ("200", "301", "302"), "{}: {}".format(status, explanation)

    headers = {}
    while True:
        line = response.readline()
        if line == "\r\n":
            break
        header, value = line.split(":", 1)
        headers[header.lower()] = value.strip()

    global redirects_times
    if status in ("301", "302"):
        if redirects_times > 10:
            raise Exception("redirects times > 10 !")
        if not headers['location']:
            raise Exception("no headers.location")
        redirects_times = redirects_times + 1
        return request(headers["location"])
    else:
        redirects_times = 0

    body = response.read()
    s.close()

    return headers, body

class Text:
    def __init__(self, text):
        self.text = text

    def __repr__(self):
        return "Text('{}'')".format(self.text)

class Tag:
    def __init__(self, tag):
        self.tag = tag

    def __repr__(self):
        return "Tag('{}')".format(self.tag)

def show(body):
    out = []
    text = ""
    in_tag = False
    for c in body:
        if c == "<":
            in_tag = True
            if text: out.append(Text(text))
            text = ""
        elif c == ">":
            in_tag = False
            out.append(Tag(text))
            text = ""
        else:
            text += c
    if not in_tag and text:
        out.append(Text(text))

    in_body = False
    body_text = ""
    for t in out:
        if isinstance(t, Text):
            if in_body:
                body_text += t.text
        elif t.tag == "body":
            in_body = True
        elif t.tag == "/body":
            in_body = False
    print(body_text)

def load(url):
    headers, body = request(url)
    show(body)


if __name__ == "__main__":
    import sys
    load(sys.argv[1])
